pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/AutoRebalancePartitionAssignmentGenerator.java

```diff
@@ -0,0 +1,113 @@
+import com.linkedin.pinot.common.utils.CommonConstants.Helix.DataSource.RoutingTableBuilderName;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import org.apache.helix.ZNRecord;
+import org.apache.helix.controller.rebalancer.strategy.AutoRebalanceStrategy;
+
+
+/**
+ * Generates partition assignment for all tables provided, using AutoRebalanceStrategy
+ */
+public class AutoRebalancePartitionAssignmentGenerator extends PartitionAssignmentGenerator {
+
+  public AutoRebalancePartitionAssignmentGenerator(TableConfig tableConfig, int nKafkaPartitions,
+      List<String> instanceNames, Map<String, TableConfig> allTableConfigsInTenant,
+      Map<String, ZNRecord> currentPartitionAssignment) {
+
+    super(tableConfig, nKafkaPartitions, instanceNames, currentPartitionAssignment);
+
+    _tablesForPartitionAssignment = new ArrayList<>();
+    for (TableConfig config : allTableConfigsInTenant.values()) {
+      RoutingConfig routingConfig = config.getRoutingConfig();
+      if (routingConfig == null || routingConfig.getRoutingTableBuilderName() == null ||
+          !RoutingTableBuilderName.PartitionAwareRealtime.toString().equals(routingConfig.getRoutingTableBuilderName())) {
+        _tablesForPartitionAssignment.add(config.getTableName());
+      }
+    }
+  }
+
+  @Override
+  public Map<String, ZNRecord> generatePartitionAssignment() {
+    Map<String, ZNRecord> tableNameToPartitionAssignment = new HashMap<>(_tablesForPartitionAssignment.size());
+    final String partitionJoiner = "_";
+
+    Map<String, Map<String, String>> currentPartitions = new HashMap<>();
+    List<String> newPartitions = new ArrayList<>();
+    for (String realtimeTableName : _tablesForPartitionAssignment) {
+
+      // construct current partitions map
+      ZNRecord partitionAssignmentZNode = _currentPartitionAssignment.get(realtimeTableName);
+      if (partitionAssignmentZNode != null) {
+
+        Map<String, List<String>> partitionToInstances = partitionAssignmentZNode.getListFields();
+        for (Map.Entry<String, List<String>> partitionEntry : partitionToInstances.entrySet()) {
+          String partitionNumber = partitionEntry.getKey();
+          List<String> partitionInstances = partitionEntry.getValue();
+          String key = realtimeTableName + partitionJoiner + partitionNumber;
+          Map<String, String> value = new HashMap<>();
+          for (String instance : partitionInstances) {
+            value.put(instance, "ONLINE");
+          }
+          currentPartitions.put(key, value);
+        }
+      }
+
+      // construct new partitions list
+      int p = _nPartitions.get(realtimeTableName);
+      for (int i = 0; i < p; i++) {
+        newPartitions.add(realtimeTableName + partitionJoiner + Integer.toString(i));
+      }
+    }
+
+    // get states
+    LinkedHashMap<String, Integer> states = new LinkedHashMap<>(2);
+    states.put("OFFLINE", 0);
```

Comment:
constants for states, they are probably defined some where else

Reply:
I see the constants defined only in CommonConstants.Helix.StateModel.SegmentOnlineOfflineStateModel.ONLINE, and did not think using that made sense here